### PR TITLE
Medical vueui updates + colours for some medical equipment

### PR DESCRIPTION
**Pull request by mikomyazaki**
Made an attempt to implement something based on this suggestion - [Brainmed UI and accessibility](https://forums.aurorastation.org/topic/13659-brainmed-ui-and-accessibility-pt-1-of-2/)

New VueUI for bodyscanner.
IPCs display their cell charge for the suit sensors program.
Handheld medical scanners have some colour coding now.

New bodyscanner UI. Bits only display when necessary, has all the information of the current nanoUI.
bodyscanner_8

### CODE REVIEW

**Comment by mikomyazaki** [comment]
!review

**Comment by alsoandanswer** [comment]
!review

**Comment by alsoandanswer** [comment]
wtf borealis is broken

**Comment by alsoandanswer** [comment]
hm, not included in the original comment, can you show the colors of handheld sensors, and infected body parts? not just necrotic.

**Comment by Geevies** [comment]
quick question, untag shells like to pretend to be humans, does this change out them as IPCs immediately? ideally untagged shells would have some sort of faked heartrate

**Comment by mikomyazaki** [comment]
> quick question, untag shells like to pretend to be humans, does this change out them as IPCs immediately? ideally untagged shells would have some sort of faked heartrate

Ummmmm yes.

Guess I need to figure out how that works, now.

**Comment by mikomyazaki** [comment]
> 
> 
> hm, not included in the original comment, can you show the colors of handheld sensors, and infected body parts? not just necrotic.

bodyscanner_9

handscanner_1

**Comment by mikomyazaki** [comment]
Now checks for tags on the ipcs before reporting a cell %, so untagged IPCs won't be outed.

**Comment by Arrow768** [comment]
Eh, I disagree with the notion that a advanced medical scanner should be unable to detect a robot placed inside of it.

The IPC already has the option to remain undetected, by avoiding advanced scanners.

**Comment by mikomyazaki** [comment]
> Eh, I disagree with the notion that a body scanner should be unable to detect a robot placed inside of it.
> 
> The IPC already has the option to remain undetected, by avoiding getting placed inside of a body scanner.

That bit is about the suit sensors not the body scanner @Arrow768

**Comment by alsoandanswer** [comment]
Find it weird that **MAJOR SYSTEMIC ORGAN DAMAGE** is in green, lol
Shouldn't it be red?
Green = Good, Yellow = Decent, Red = Bad

**Comment by mikomyazaki** [comment]
> Find it weird that **MAJOR SYSTEMIC ORGAN DAMAGE** is in green, lol
> Shouldn't it be red?
> Green = Good, Yellow = Decent, Red = Bad

Colours are hard.

Thank you, fixed...

**Comment by alsoandanswer** [comment]
Screenshot_20200126-134504_Chrome

checked travis, not sure if related but should look at it

**Comment by mikomyazaki** [comment]
Don't PR when you're exhausted @_@

**Comment by mikomyazaki** [comment]
Needs some vueui tweaks before rereview

**Comment by mikomyazaki** [comment]
!review

File: code/game/machinery/adv_med.dm [hunk 1]
```diff
@@ -355,46 +362,40 @@
...
-		data["hastgvirus"]		= occupant.viruses.len
-		data["tgvirus"]			= occupant.viruses
-
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, data, force_open)
-	if (!ui)
-		ui = new(user, src, ui_key, "med_diagnostics.tmpl", "Medical Diagnostics", 800, 500, state = state)
-		ui.set_initial_data(data)
-		ui.open()
-		ui.set_auto_update(1)
+		VUEUI_SET_CHECK(data["missingparts"], missing, ., data)
```
Comment (Karolis2011) [inline code/game/machinery/adv_med.dm]: This is absolute flop. List instances always will differ it UI pushes change. You can't do this with set check.

File: code/game/machinery/adv_med.dm [hunk 2]
```diff
@@ -355,46 +362,40 @@
...
+		VUEUI_SET_CHECK(data["paralysis"], occupant.paralysis, ., data)
+		VUEUI_SET_CHECK(data["bodytemp"], occupant.bodytemperature, ., data)
+		VUEUI_SET_CHECK(data["occupant"], occupant, ., data)
+		VUEUI_SET_CHECK(data["norepiAmt"], R.get_reagent_amount("norepinephrine"), ., data)
+		VUEUI_SET_CHECK(data["soporAmt"], R.get_reagent_amount("stoxin"), ., data)
+		VUEUI_SET_CHECK(data["bicardAmt"], R.get_reagent_amount("bicaridine"), ., data)
+		VUEUI_SET_CHECK(data["dexAmt"], R.get_reagent_amount("dexalin"), ., data)
+		VUEUI_SET_CHECK(data["dermAmt"], R.get_reagent_amount("dermaline"), ., data)
+		VUEUI_SET_CHECK(data["otherAmt"], R.total_volume - (data["soporAmt"] + data["dexAmt"] + data["bicardAmt"] + data["norepiAmt"] + data["dermAmt"]), ., data)
+		VUEUI_SET_CHECK(data["bodyparts"], get_external_wound_data(occupant), ., data)
```
Comment (Karolis2011) [inline code/game/machinery/adv_med.dm]: I guess this also returns list. What is also a no no.

File: code/game/machinery/adv_med.dm [hunk 3]
```diff
@@ -355,46 +362,40 @@
...
+
+		VUEUI_SET_CHECK(data["cloneLoss"], get_severity(occupant.getCloneLoss()), ., data)
+		VUEUI_SET_CHECK(data["oxyLoss"], get_severity(occupant.getOxyLoss()), ., data)
+		VUEUI_SET_CHECK(data["bruteLoss"], get_severity(occupant.getBruteLoss()), ., data)
+		VUEUI_SET_CHECK(data["fireLoss"], get_severity(occupant.getFireLoss()), ., data)
+		VUEUI_SET_CHECK(data["toxLoss"], get_severity(occupant.getToxLoss()), ., data)
+
+		VUEUI_SET_CHECK(data["paralysis"], occupant.paralysis, ., data)
+		VUEUI_SET_CHECK(data["bodytemp"], occupant.bodytemperature, ., data)
+		VUEUI_SET_CHECK(data["occupant"], occupant, ., data)
```
Comment (Karolis2011) [inline code/game/machinery/adv_med.dm]: _Are you passing here a datum to UI?_ what is this heresy

File: code/game/machinery/adv_med.dm [hunk 4]
```diff
@@ -314,29 +313,37 @@
...
-
-	data["noscan"]		= src.connected.check_species()
-	data["nocons"]		= !src.connected
-	data["occupied"] 	= occupied
-	data["invalid"]		= src.connected && src.connected.check_species()
-	data["ipc"]			= src.connected && occupant && isipc(occupant)
+	if (connected)
+		occupant = connected.occupant
+
+		VUEUI_SET_CHECK(data["noscan"], !!connected.check_species(), ., data)
```
Comment (Karolis2011) [inline code/game/machinery/adv_med.dm]: Conditional field setting is a no no. All fields must be present at all times. UI will faill to update if extra root data fields are added afterwards.

File: code/game/machinery/adv_med.dm [hunk 5]
```diff
@@ -355,46 +362,40 @@
...
-		data["tgvirus"]			= occupant.viruses
-
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, data, force_open)
-	if (!ui)
-		ui = new(user, src, ui_key, "med_diagnostics.tmpl", "Medical Diagnostics", 800, 500, state = state)
-		ui.set_initial_data(data)
-		ui.open()
-		ui.set_auto_update(1)
+		VUEUI_SET_CHECK(data["missingparts"], missing, ., data)
+		VUEUI_SET_CHECK(data["hasmissing"], !!missing.len, ., data)
```
Comment (Karolis2011) [inline code/game/machinery/adv_med.dm]: Here you can just pass Len, no need for `!!`

File: code/game/machinery/adv_med.dm [hunk 6]
```diff
@@ -355,46 +362,40 @@
...
+		VUEUI_SET_CHECK(data["hasvirus"], occupant.virus2.len || occupant.viruses.len, ., data)
+		VUEUI_SET_CHECK(data["hastgvirus"], occupant.viruses.len, ., data)
+		VUEUI_SET_CHECK(data["tgvirus"], occupant.viruses, ., data)
```
Comment (Karolis2011) [inline code/game/machinery/adv_med.dm]: Again, list directly in Vue set check

File: code/modules/mob/living/carbon/human/human.dm [hunk 7]
```diff
@@ -1810,11 +1810,50 @@
...
+	switch(bp[2])
+		if(BP_HIGH_DIASTOLIC to INFINITY)
+			diastolic_alert = 4
+		if(BP_PRE_HIGH_DIASTOLIC to BP_HIGH_DIASTOLIC)
+			diastolic_alert = 3
+		if(BP_IDEAL_DIASTOLIC to BP_PRE_HIGH_DIASTOLIC)
+			diastolic_alert = 2
+		if(-INFINITY to BP_IDEAL_DIASTOLIC)
+			diastolic_alert = 1
+	if(systolic_alert == 4 || diastolic_alert == 4)
```
Comment (Karolis2011) [inline code/modules/mob/living/carbon/human/human.dm]: Magic numbers.

File: vueui/src/components/view/mcomputer/medical/sensors.vue [hunk 8]
```diff
@@ -83,6 +84,26 @@ export default {
...
+        case 3:
+          return "average"
+        case 4:
+          return "bad"
+        default:
+          return "neutral"
+      }
+    },
+    getChargeClass(cellCharge) {
+        if(cellCharge > 10) {
```
Comment (Karolis2011) [inline vueui/src/components/view/mcomputer/medical/sensors.vue]: What is this indentation. It's a mess.

File: vueui/src/components/view/medical/bodyscanner.vue [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    <div v-if="invalid" class="center">
+      <span class="red" v-if="nocons">Error: No scanner bed detected!</span>
+      <span class="red" v-else-if="!occupied">No occupant detected.</span>
+      <span class="red" v-else-if="ipc">Error: Object in scanner bed interfering with sensor array.</span>
+      <span class="red" v-else-if="noscan">Error: No diagnostics profile installed for this species.</span>
+      <span class="red" v-else>Error: Unknown error.</span>
+    </div>
+
+    <!-- Health Display -->
+    <div class="table" v-else>
```
Comment (Karolis2011) [inline vueui/src/components/view/medical/bodyscanner.vue]: Please use `<table>`

File: vueui/src/components/view/medical/bodyscanner.vue [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+
+
+<script>
+  export default {
+    data() {
+      return this.$root.$data.state;
+    },
+    methods: {
+    round(value) {
+      return Math.round(value)
```
Comment (Karolis2011) [inline vueui/src/components/view/medical/bodyscanner.vue]: What is this indentation. Also why you couldn't use `Math.round()` directly in template?

**Pull request merged**